zanata-war/src/main/java/org/zanata/service/impl/MergeTranslationsWork.java

```diff
@@ -0,0 +1,137 @@
+package org.zanata.service.impl;
+
+import java.util.List;
+
+import lombok.AllArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+import org.jboss.seam.core.Events;
+import org.jboss.seam.util.Work;
+import org.zanata.common.ContentState;
+import org.zanata.dao.TextFlowDAO;
+import org.zanata.events.TextFlowTargetStateEvent;
+import org.zanata.model.HLocale;
+import org.zanata.model.HSimpleComment;
+import org.zanata.model.HTextFlow;
+import org.zanata.model.HTextFlowTarget;
+import org.zanata.service.TranslationStateCache;
+import org.zanata.util.MessageGenerator;
+
+import com.google.common.base.Stopwatch;
+
+/**
+ * Merge translation and persist in transaction.
+ *
+ * Merge HTextFlowTargets from HProjectIteration(id=sourceVersionId) to
+ * HProjectIteration(id=targetVersionId) in batches(batchStart, batchLength)
+ *
+ * @see org.zanata.service.impl.MergeTranslationsServiceImpl#startMergeTranslations
+ * @see org.zanata.service.impl.MergeTranslationsServiceImpl#shouldMerge
+ *
+ * returns count of processed translations.
+ *
+ * @author Alex Eng <a href="mailto:aeng@redhat.com">aeng@redhat.com</a>
+ */
+@Slf4j
+@AllArgsConstructor
+public class MergeTranslationsWork extends Work<Integer> {
+    private final Long sourceVersionId;
+    private final Long targetVersionId;
+    private final int batchStart;
+    private final int batchLength;
+    private final boolean useNewerTranslation;
+    private final List<HLocale> supportedLocales;
+
+    private final TextFlowDAO textFlowDAO;
+
+    private final TranslationStateCache translationStateCacheImpl;
+
+    private final Stopwatch stopwatch = Stopwatch.createUnstarted();
+
+    @Override
+    protected Integer work() throws Exception {
+        stopwatch.start();
+
+        List<HTextFlow[]> matches =
+                textFlowDAO.getSourceByMatchedContext(
+                        sourceVersionId, targetVersionId, batchStart,
+                        batchLength);
+
+        for (HTextFlow[] results : matches) {
+            HTextFlow sourceTf = results[0];
+            HTextFlow targetTf = results[1];
+            boolean hasChanged = false;
```

Comment:
I would call this `foundChange` so that it is always true. We know right here that we did not find a change yet so we can honestly set it to `false`, and we know that if we get to the end and we have not found a change, we can assume there are no changes.